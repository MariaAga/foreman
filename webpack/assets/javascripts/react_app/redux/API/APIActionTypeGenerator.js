const actions = ['REQUEST', 'SUCCESS', 'FAILURE'];

/**
 * Auto generates api consts for redux for given subtype
 * @param {String} subtype - the name of the component
 * @param {Object} actionTypes - custom types to use instead of the auto generated ones
 */
export const actionTypeGenerator = (subtype, actionTypes = {}) => {
  actions.forEach(type => {
    actionTypes[type] = actionTypes[type] || `${subtype}_${type}`;
  });
  return actionTypes;
};
