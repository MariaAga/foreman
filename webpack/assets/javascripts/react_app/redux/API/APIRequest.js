import { API } from './';

export const get = async (
  payload,
  url,
  store,
  actionTypes,
  errorFormat,
  successFormat,
  onSuccess,
  onFailure
) => {
  store.dispatch({
    type: actionTypes.REQUEST,
    payload,
  });
  try {
    const response = await API.get(
      url,
      payload.headers || {},
      payload.params || {}
    );
    onSuccess(response);
    store.dispatch({
      type: actionTypes.SUCCESS,
      payload: { ...payload, ...successFormat(response.data) },
    });
  } catch (error) {
    onFailure(error);
    store.dispatch({
      type: actionTypes.FAILURE,
      payload: { ...errorFormat({ error }), payload },
    });
  }
};
