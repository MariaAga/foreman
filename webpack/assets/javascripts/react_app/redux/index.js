import createLogger from 'redux-logger';
import thunk from 'redux-thunk';
import { applyMiddleware, createStore } from 'redux';

import reducers from './reducers';

import { ajaxRequestAction } from './middlewares/ajaxRequestAction';

let middleware = [thunk, ajaxRequestAction];

const useLogger = () => {
  const isProduction = process.env.NODE_ENV === 'production';
  const isLogger = process.env.REDUX_LOGGER;

  if (!isProduction && !global.__testing__) {
    if (isLogger === undefined || isLogger === true) return true;
  }
  return isProduction && isLogger;
};

if (useLogger()) middleware = [...middleware, createLogger()];

export const generateStore = () =>
  createStore(
    reducers,
    window.__REDUX_DEVTOOLS_EXTENSION__ &&
      window.__REDUX_DEVTOOLS_EXTENSION__(),
    applyMiddleware(...middleware)
  );

const store = generateStore();

export default store;
