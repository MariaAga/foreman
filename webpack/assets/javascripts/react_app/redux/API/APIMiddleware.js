import { API_OPERATIONS, actionTypeGenerator } from './';
import { get } from './APIRequest';

export const APIMiddleware = store => next => action => {
  const { type, outputType, payload = {}, url, actionTypes = {} } = action;
  if (type === API_OPERATIONS.GET) {
    get(payload, url, store, actionTypeGenerator(outputType, actionTypes));
  } else if (type === '_START_POLLING') {
    pollingStart(
      payload.timeoutId,
      outputType,
      payload.APIParams,
      payload.interval
    );
  } else if (type === '_STOP_POLLING') {
    pollingStart(
      payload.timeoutId,
      outputType,
      payload.APIParams,
      payload.interval
    );
  } else {
    next(action);
  }
};

export const pollingStop = (timeoutId, outputType) => ({
  type: `${outputType}_STOP_POLLING`,
  payload: { timeoutId, outputType },
});

export const pollingStart = (
  timeoutId,
  outputType,
  APIParams,
  interval = 5000
) => dispatch => {
  if (!timeoutId) {
    timeoutId = setInterval(() => getData(APIParams, outputType), interval);
  }
  dispatch({
    type: `${outputType}_START_POLLING`,
    payload: { timeoutId },
  });
};

const getData = (APIParams, outputType) => ({
  type: API_OPERATIONS.GET,
  outputType,
  ...APIParams,
});
