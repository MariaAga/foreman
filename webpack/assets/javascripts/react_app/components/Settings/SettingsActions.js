import { GET_SETTING } from './SettingsConstants';
import { API_OPERATIONS } from '../../redux/API';

export const loadSetting = settingName => dispatch => {
  dispatch({
    type: API_OPERATIONS.GET,
    outputType: GET_SETTING,
    url: `/api/v2/settings/${settingName}`,
    successFormat: data => ({ data }),
  });
};

export default loadSetting;
