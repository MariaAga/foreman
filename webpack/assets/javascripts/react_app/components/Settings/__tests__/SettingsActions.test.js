import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import API from '../../../redux/API/API';
import { loadSetting } from '../SettingsActions';
import { APIMiddleware } from '../../../redux/API';
import IntegrationTestHelper from '../../../common/IntegrationTestHelper';

jest.mock('../../../redux/API/API');
const middlewares = [thunk, APIMiddleware];
const mockStore = configureMockStore(middlewares);
const store = mockStore();

const successResponse = {
  data: 'some-data',
};

afterEach(() => {
  store.clearActions();
});

describe('Settings actions', () => {
  it('should load settings and success', async () => {
    API.get.mockImplementation(
      url =>
        new Promise((resolve, reject) => {
          resolve(successResponse);
        })
    );
    await store.dispatch(loadSetting('some-name'));
    await IntegrationTestHelper.flushAllPromises();
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should load settings and fail', async () => {
    API.get.mockImplementation(
      url =>
        new Promise((resolve, reject) => {
          reject(Error('some-error'));
        })
    );
    await store.dispatch(loadSetting('some-name'));
    await IntegrationTestHelper.flushAllPromises();
    expect(store.getActions()).toMatchSnapshot();
  });
});
