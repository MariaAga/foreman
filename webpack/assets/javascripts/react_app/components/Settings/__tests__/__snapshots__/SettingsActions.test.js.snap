// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings actions should load settings and fail 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "GET_SETTING_REQUEST",
  },
  Object {
    "payload": Object {
      "error": [Error: some-error],
      "payload": Object {},
    },
    "type": "GET_SETTING_FAILURE",
  },
]
`;

exports[`Settings actions should load settings and success 1`] = `
Array [
  Object {
    "payload": Object {},
    "type": "GET_SETTING_REQUEST",
  },
  Object {
    "payload": Object {
      "data": "some-data",
    },
    "type": "GET_SETTING_SUCCESS",
  },
]
`;
